### PR TITLE
[SPARK-38138][SQL] Materialize QueryPlan subqueries

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

This PR propose to materialize `QueryPlan#subqueries` and pruned by `PLAN_EXPRESSION` on searching to improve the SQL compile performance.

### Why are the changes needed?

We found a query in production that cost lots of time in optimize phase (also include AQE optimize phase) when enable DPP, the SQL pattern likes

```
select <cols...>
from a
left join b on a.<col> = b.<col>
left join c on b.<col> = c.<col>
left join d on c.<col> = d.<col>
left join e on d.<col> = e.<col>
left join f on e.<col> = f.<col>
left join g on f.<col> = g.<col>
left join h on g.<col> = h.<col>
...
```
SPARK-36444 significantly reduces the optimize time (exclude AQE phase), see detail at #35431, but there are still lots of time costs in `InsertAdaptiveSparkPlan` on AQE optimize phase.

Before this change, the query costs 658s, after this change only costs 65s.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing UTs.

### CODE REVIEW

**Comment by pan3793** [comment]
cc @wangyum @cloud-fan @yaooqinn

**Comment by HyukjinKwon** [comment]
cc @maryannxue @allisonwang-db @sigmod FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala [hunk 1]
```diff
@@ -427,8 +427,8 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]]
   /**
    * All the top-level subqueries of the current plan node. Nested subqueries are not included.
    */
-  def subqueries: Seq[PlanType] = {
-    expressions.flatMap(_.collect {
+  lazy val subqueries: Seq[PlanType] = {
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala:430]: please add `@transient`

Comment (pan3793) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: Thanks for tips, updated.

**Comment by amaliujia** [comment]
cc @amaliujia

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala (hunk 1 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: Just for education purpose: why `@transient` is useful here?

Comment (pan3793) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: The `QueryPlan` e.g. `Expression` need to be sent to executor, use `@transient` to reduce the memory usage of executor.

**Comment by pan3793** [comment]
@cloud-fan would you please take a look? thanks

**Review by sigmod — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**